File: desktop/capture/swift/scroll-capture.swift
// desktop/capture/swift/scroll-capture.swift
//
// Generic scrolling capture helper. Snaps the requested rectangle, sends a
// scroll-wheel event to whatever window owns that area, waits for the page
// to settle, snaps again, and repeats until either:
//   - the next frame is byte-identical to the previous (page bottom)
//   - the requested max number of scrolls is reached
//
// Frames are written as `frame-NNNN.png` into the output directory. Stitching
// is delegated to scroll-stitch.swift — separating the two means we can
// debug capture artefacts independently from the Vision-framework alignment
// pass.
//
// Args (positional, all required):
//   x y w h   — capture rectangle in CSS pixels (Cocoa flipped Y)
//   outDir    — directory to receive frame-NNNN.png
//   scrolls   — hard cap on iterations
//   delayMs   — wait after each scroll before snapping the next frame
//
// Stdout: `frame <path>` per saved frame, `done <count>` on success,
//         `settled <index>` when the loop short-circuits on a duplicate.
// Stderr: human-readable error messages.

import Foundation
import CoreGraphics
import AppKit
import ScreenCaptureKit

// MARK: - Arg parsing -----------------------------------------------------

guard CommandLine.arguments.count == 8 else {
    FileHandle.standardError.write("usage: scroll-capture x y w h outDir scrolls delayMs\n".data(using: .utf8)!)
    exit(2)
}

let argv = CommandLine.arguments
guard
    let x = Double(argv[1]),
    let y = Double(argv[2]),
    let w = Double(argv[3]),
    let h = Double(argv[4]),
    let maxScrolls = Int(argv[6]),
    let delayMs = Int(argv[7])
else {
    FileHandle.standardError.write("invalid numeric argument\n".data(using: .utf8)!)
    exit(2)
}
let outDir = argv[5]
let scrollAmount: Int32 = -8   // lines per wheel tick — negative = scroll down

// MARK: - Output dir ------------------------------------------------------

do {
    try FileManager.default.createDirectory(atPath: outDir, withIntermediateDirectories: true)
} catch {
    FileHandle.standardError.write("cannot create outDir: \(error)\n".data(using: .utf8)!)
    exit(3)
}

// MARK: - Helpers ---------------------------------------------------------

func emit(_ s: String) {
    FileHandle.standardOutput.write("\(s)\n".data(using: .utf8)!)
}

func sleepMs(_ ms: Int) {
    Thread.sleep(forTimeInterval: Double(ms) / 1000.0)
}

func postScrollDown() {
    // CGEvent vertical wheel — post at HID level so the OS routes the event
    // to whichever window currently sits under the cursor.
    if let ev = CGEvent(scrollWheelEvent2Source: nil,
                       units: .line,
                       wheelCount: 1,
                       wheel1: scrollAmount,
                       wheel2: 0,
                       wheel3: 0) {
        ev.post(tap: .cghidEventTap)
    }
}

// Cheap byte-stream fingerprint — sum of every 1024th byte. Good enough to
// detect "page didn't scroll" since two identical PNG encodes yield matching
// streams. Avoids pulling in CommonCrypto for what is effectively a hash.
func fingerprint(_ data: Data) -> String {
    var sum: UInt64 = 0
    data.withUnsafeBytes { (buf: UnsafeRawBufferPointer) in
        let count = buf.count
        var i = 0
        while i < count {
            sum = sum &+ UInt64(buf[i])
            i += 1024
        }
    }
    return String(sum)
}

// MARK: - ScreenCaptureKit one-shot capture ------------------------------

// Resolve the primary display via SCShareableContent. Cache the filter so we
// only trigger the TCC prompt once.
let captureRect = CGRect(x: x, y: y, width: w, height: h)

actor Recorder {
    private var filter: SCContentFilter?
    private var configBase: SCStreamConfiguration?
    private var scale: Double = 2.0

    func prepare() async throws {
        let content = try await SCShareableContent.current
        guard let display = content.displays.first else {
            throw NSError(domain: "ScrollCapture", code: 1,
                          userInfo: [NSLocalizedDescriptionKey: "No available display"])
        }
        scale = await MainActor.run { NSScreen.main?.backingScaleFactor ?? 2 }
        filter = SCContentFilter(display: display, excludingWindows: [])
        let config = SCStreamConfiguration()
        // SourceRect crops on the display side so we don't have to resample.
        config.sourceRect = captureRect
        config.width = Int(captureRect.width * scale)
        config.height = Int(captureRect.height * scale)
        config.pixelFormat = kCVPixelFormatType_32BGRA
        config.showsCursor = false
        configBase = config
    }

    func snap() async throws -> Data {
        guard let filter, let configBase else {
            throw NSError(domain: "ScrollCapture", code: 2,
                          userInfo: [NSLocalizedDescriptionKey: "Recorder not prepared"])
        }
        let cgImage = try await SCScreenshotManager.captureImage(
            contentFilter: filter,
            configuration: configBase
        )
        let rep = NSBitmapImageRep(cgImage: cgImage)
        guard let pngData = rep.representation(using: .png, properties: [:]) else {
            throw NSError(domain: "ScrollCapture", code: 3,
                          userInfo: [NSLocalizedDescriptionKey: "PNG encode failed"])
        }
        return pngData
    }
}

// MARK: - Main async pump -------------------------------------------------

let group = DispatchGroup()
group.enter()

Task {
    defer { group.leave() }

    let recorder = Recorder()
    do {
        try await recorder.prepare()
    } catch {
        FileHandle.standardError.write("prepare failed: \(error)\n".data(using: .utf8)!)
        exit(4)
    }

    // Centre the cursor over the capture rect so scroll events route to the
    // intended window. The cursor is captured-then-moved before we ever take
    // a frame, so the first PNG doesn't show the warp.
    let centre = CGPoint(x: captureRect.midX, y: captureRect.midY)
    CGWarpMouseCursorPosition(centre)
    sleepMs(120)

    var lastFingerprint: String? = nil
    var saved = 0

    for i in 0..<maxScrolls {
        let pngData: Data
        do {
            pngData = try await recorder.snap()
        } catch {
            FileHandle.standardError.write("snap failed at \(i): \(error)\n".data(using: .utf8)!)
            exit(5)
        }
        let filename = String(format: "frame-%04d.png", i)
        let fullPath = (outDir as NSString).appendingPathComponent(filename)
        do {
            try pngData.write(to: URL(fileURLWithPath: fullPath))
        } catch {
            FileHandle.standardError.write("write failed at \(i): \(error)\n".data(using: .utf8)!)
            exit(6)
        }
        emit("frame \(fullPath)")
        saved += 1

        let print = fingerprint(pngData)
        if let prev = lastFingerprint, prev == print {
            emit("settled \(i)")
            break
        }
        lastFingerprint = print

        postScrollDown()
        sleepMs(delayMs)
    }

    emit("done \(saved)")
}

group.wait()

File: desktop/capture/scroll-capture.ts
// desktop/capture/scroll-capture.ts
//
// Orchestrates the two Swift helpers — scroll-capture (frame grabber +
// CGEvent scroll driver) and scroll-stitch (Vision-framework concatenator)
// — into a single async run. Caller picks the rectangle, this module:
//
//   1. spawns scroll-capture with the area + a hard-coded "max 30 scrolls,
//      400 ms settle delay" budget; reads frame paths from its stdout
//   2. waits for `done <N>` or `settled <i>` and collects all frames
//   3. spawns scroll-stitch with `out=...` + the collected frame paths
//   4. resolves with the path to the stitched PNG
//
// Failure modes are reported as plain Error objects so the caller can
// surface them to the user without unwrapping subprocess plumbing.

import { spawn } from "node:child_process";
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { asarUnpacked } from "../utils/asar-paths.ts";

const currentFilePath = fileURLToPath(import.meta.url);
const desktopDistDir = path.dirname(currentFilePath);
// asarUnpacked() — `child_process.spawn` cannot descend into app.asar (#82).
const desktopDistDirOnDisk = asarUnpacked(desktopDistDir);
const scrollCaptureBin = path.join(desktopDistDirOnDisk, "scroll-capture");
const scrollStitchBin = path.join(desktopDistDirOnDisk, "scroll-stitch");

export interface ScrollCaptureOptions {
  /** Capture rectangle in CSS pixels. */
  area: { x: number; y: number; width: number; height: number };
  /** Hard cap on iterations. Default: 30. */
  maxScrolls?: number;
  /** Settle delay between scrolls in ms. Default: 400. */
  delayMs?: number;
  /** Optional path for the final stitched PNG. Defaults to a tmp file. */
  outPath?: string;
}

export interface ScrollCaptureResult {
  outPath: string;
  frameCount: number;
  settledEarly: boolean;
}

export class ScrollCapture {
  static async isAvailable(): Promise<boolean> {
    try {
      await fs.access(scrollCaptureBin);
      await fs.access(scrollStitchBin);
      return true;
    } catch {
      return false;
    }
  }

  async run(opts: ScrollCaptureOptions): Promise<ScrollCaptureResult> {
    if (!(await ScrollCapture.isAvailable())) {
      throw new Error(
        "Scrolling capture is unavailable — Swift helpers were not compiled. " +
          "Re-run `npm run build` on macOS to regenerate them."
      );
    }

    const maxScrolls = opts.maxScrolls ?? 30;
    const delayMs = opts.delayMs ?? 400;
    const framesDir = await fs.mkdtemp(path.join(os.tmpdir(), "marshal-scroll-"));
    const outPath =
      opts.outPath ??
      path.join(framesDir, `scroll-${Date.now()}.png`);

    // Step 1 — capture frames.
    const { frames, settledEarly } = await this.runFrameCapture({
      area: opts.area,
      framesDir,
      maxScrolls,
      delayMs
    });

    if (frames.length === 0) {
      await fs.rm(framesDir, { recursive: true, force: true }).catch(() => undefined);
      throw new Error("scroll-capture produced no frames");
    }

    // Step 2 — stitch.
    try {
      await this.runStitch(frames, outPath);
    } catch (err) {
      // Keep the frames around for debugging if stitching fails — saves the
      // user a re-capture and gives us something to file with the bug.
      throw new Error(
        `Stitching failed (${(err as Error).message}). ` +
          `Raw frames left in ${framesDir} for inspection.`
      );
    }

    // Best-effort cleanup of the individual frames; stitched PNG already
    // copied out by scroll-stitch.
    if (!outPath.startsWith(framesDir)) {
      await fs.rm(framesDir, { recursive: true, force: true }).catch(() => undefined);
    }

    return {
      outPath,
      frameCount: frames.length,
      settledEarly
    };
  }

  private runFrameCapture(input: {
    area: ScrollCaptureOptions["area"];
    framesDir: string;
    maxScrolls: number;
    delayMs: number;
  }): Promise<{ frames: string[]; settledEarly: boolean }> {
    return new Promise((resolve, reject) => {
      const args = [
        String(input.area.x),
        String(input.area.y),
        String(input.area.width),
        String(input.area.height),
        input.framesDir,
        String(input.maxScrolls),
        String(input.delayMs)
      ];
      const child = spawn(scrollCaptureBin, args, { stdio: ["ignore", "pipe", "pipe"] });

      const frames: string[] = [];
      let settledEarly = false;
      let stderr = "";
      let stdoutBuf = "";

      child.stdout.on("data", (chunk: Buffer) => {
        stdoutBuf += chunk.toString("utf8");
        let nlIndex: number;
        while ((nlIndex = stdoutBuf.indexOf("\n")) >= 0) {
          const line = stdoutBuf.slice(0, nlIndex).trim();
          stdoutBuf = stdoutBuf.slice(nlIndex + 1);
          if (line.startsWith("frame ")) {
            frames.push(line.slice("frame ".length));
          } else if (line.startsWith("settled ")) {
            settledEarly = true;
          }
          // `done <N>` arrives just before the process exits; we don't need to
          // parse it — the exit handler resolves regardless.
        }
      });

      child.stderr.on("data", (chunk: Buffer) => {
        stderr += chunk.toString("utf8");
      });

      child.on("error", (err) => reject(err));
      child.on("exit", (code) => {
        if (code === 0) {
          resolve({ frames, settledEarly });
        } else {
          reject(new Error(`scroll-capture exited ${code}: ${stderr.trim() || "no stderr"}`));
        }
      });
    });
  }

  private runStitch(frames: string[], outPath: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const args = [`out=${outPath}`, ...frames];
      const child = spawn(scrollStitchBin, args, { stdio: ["ignore", "pipe", "pipe"] });

      let stderr = "";
      child.stderr.on("data", (chunk: Buffer) => {
        stderr += chunk.toString("utf8");
      });

      child.on("error", (err) => reject(err));
      child.on("exit", (code) => {
        if (code === 0) {
          resolve();
        } else {
          reject(new Error(`scroll-stitch exited ${code}: ${stderr.trim() || "no stderr"}`));
        }
      });
    });
  }
}
